go/consensus/api/transaction/v0/envelope.go

```diff
@@ -0,0 +1,201 @@
+// Package v0 implements the version 0 transaction envelope format and
+// associated types.
+package v0
+
+import (
+	"fmt"
+	"math/bits"
+
+	"github.com/oasisprotocol/oasis-core/go/common/cbor"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+	"github.com/oasisprotocol/oasis-core/go/common/crypto/signature"
+)
+
+// AccountSigner is a signer associated with an account.
+type AccountSigner struct {
+	// PublicKey is the account signer's public key.
+	PublicKey signature.PublicKey `json:"public_key"`
+
+	// Weight is the account signer's signing weight.
+	Weight uint64 `json:"weight"`
+}
+
+// Account is an account descriptor.
+type Account struct {
+	cbor.Versioned
+
+	// Signers are the account signers associated with the given
+	// account.
+	Signers []AccountSigner `json:"signers,omitempty"`
+
+	// Threshold is the minimum combined weight that must be
+	// met (>=) for // a transaction to be considered valid.
+	Threshold uint64 `json:"threshold"`
+}
+
+// Verify validates an account descriptor for well-formedness.
+func (acc *Account) Verify() error {
+	if acc.V != 0 {
+		return fmt.Errorf("consensus/transaction/v0: invalid version: %v", acc.V)
+	}
+	if len(acc.Signers) == 0 {
+		return fmt.Errorf("consensus/transaction/v0: no account signers")
+	}
+
```

Comment:
Should we add a maximum number of allowed signers? Should additional signers cost extra gas (beyond just the txes being larger) or is verifying extra signatures not really that more expensive since we're doing batch verification?

Reply:
This "depends".  If the number of signatures is larger than the internal/unexposed minimum batch size, and all the signatures are correct, then it's fast.  Otherwise it validates the signatures serially.

This does mean the cases where this actually can be slow, there's no one to charge (any one signature invalidates the envelope), so this is probably ok for now.